[win] Adding experimental preference to enable update on runtime behavior

ShahzaibIbrahim opened the pull request:
In the UI Preferences -> General -> Appearance, there is a new HiDPI setting that could be checked to enable the monitor-specific scaling of the UI. This check will also enable Edge browser by default. This is an experimental feature to test the multi-monitor HiDPI support before it is actually released and set by default. For now, this preference is disabled by default.

**Note: This preference check is only available in Windows** 

![image](https://github.com/user-attachments/assets/a1288374-5aeb-436f-9c8e-7557863f12e2)

When changing the preference, the dialog will appear to prompt the restart. 

![image](https://github.com/user-attachments/assets/8580a8fd-73f8-42c1-b79f-22ed031fe04c)

laeubi commented:
As we have ways to know the platform, maybe only show it when running on windows?

Also either make PerMonitorV2 a link to some documentation or remove implementation details from the description.

BeckerWdf commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -192,6 +200,32 @@ protected Control createContents(Composite parent) {
+		buttonComposite.setLayout(layout);
+		Label infoLabel = new Label(buttonComposite, SWT.WRAP);
+		infoLabel.setText(
+				"This preference is Windows only. Activating it will change the DPI awareness of the IDE to PerMonitorV2 and dynamically scale all windows according to the monitor it is currently assigned to. An additional effect of this preference is to set the default browser to Edge to provide the appropriate scaling of content displayed in a browser. This feature is still in development and therefore mar ...
```
Shouldn't we move this text to a messages.properties file?

I would write:
"This setting is specific to the Windows operating system..."

ShahzaibIbrahim replied:
The only reason I didn't bother to put the message in the file is that later it would be easier to remove this temporary option (just from one place) but yeah  I will move it.

ShahzaibIbrahim commented:
> As we have ways to know the platform, maybe only show it when running on windows?

@laeubi Forgot to mention but yes, we are only showing it in Windows. I have updated the description.

laeubi commented:
If it is only shown on windows, then I don't think it needs to be mentioned as a "Windows only" in the text.

BeckerWdf commented:
> Also either make PerMonitorV2 a link to some documentation or remove implementation details from the description.

One could also add such details into the documentation page for this preference page.

ShahzaibIbrahim commented:
@BeckerWdf @laeubi for now I have removed implementation details from the description. I have asked if we have documentation page for this right now. If yes, then I will link it.

fedejeanne commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/Workbench.java:
```diff
@@ -656,6 +656,17 @@ public void update() {
+		if (updateOnRunTime) {
+			System.setProperty("org.eclipse.swt.browser.DefaultType", "edge"); //$NON-NLS-1$ //$NON-NLS-2$
+		} else {
+			System.setProperty("org.eclipse.swt.browser.DefaultType", "ie"); //$NON-NLS-1$ //$NON-NLS-2$
```
By explicitly setting this to `ie` you are also overriding the value of `-Dorg.eclipse.swt.browser.DefaultType` that might have been set from outside. I would remove the `else` and let the user select whichever browser s/he wants if "update on runtime" is not activated.

fedejeanne commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -114,6 +116,9 @@ public class ViewsPreferencePage extends PreferencePage implements IWorkbenchPre
 
 	private Button themingEnabled;
+	private Button updateOnRunTime;
+	boolean initialStateUpdateOnRuntime = PrefUtil.getAPIPreferenceStore()
```
make it `private`

fedejeanne commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -192,6 +200,31 @@ protected Control createContents(Composite parent) {
 		return comp;
 	}
 
+	private void createUpdateOnRuntimeButton(Composite composite) {
```
```suggestion
	private void createHiDPISettingsGroup(Composite parent) {
```

fedejeanne commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -192,6 +200,31 @@ protected Control createContents(Composite parent) {
+			return;
+		}
+		createLabel(composite, ""); //$NON-NLS-1$
+		Group buttonComposite = new Group(composite, SWT.LEFT);
```
```suggestion
		Group group = new Group(composite, SWT.LEFT);
```

fedejeanne commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -341,6 +373,11 @@ public boolean performOk() {
 				.getSelection();
 		prefs.putBoolean(PartRenderingEngine.ENABLED_THEME_KEY, themingEnabled.getSelection());
 
+		boolean isUpdateOnRuntTmeChanged = (updateOnRunTime != null
```
fix 2 typos
```suggestion
		boolean isUpdateOnRuntimeChanged = (updateOnRuntime != null
```

fedejeanne commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -373,13 +410,19 @@ public boolean performOk() {
 			showRestartDialog();
 		}
+		if (isUpdateOnRuntTmeChanged) {
+			showRestartDialog("DPI Setting Changed", "Restart for the effect to take place"); //$NON-NLS-1$ //$NON-NLS-2$
```
Externalize these strings.

fedejeanne commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/messages.properties:
```diff
@@ -497,6 +497,8 @@ PreferencePageParameterValues_pageLabelSeparator = \ >\
 ThemeChangeWarningText = Restart for the theme changes to take full effect
 ThemeChangeWarningTitle = Theme Changed
+UpdateUIZoomTitle = Update UI scaling on runtime
+UpdateUIZoomDesc = Activating it will dynamically scale all windows according to the monitor it is currently assigned to. An additional effect of this preference is to set the default browser to Edge to provide the appropriate scaling of content displayed in a browser. This feature is still in development and therefore marked as experimental.
```
```suggestion
UpdateUIZoomDesc = EXPERIMENTAL! Activating this option will dynamically scale all windows according to the monitor they are currently in. It will also set the default browser to Edge in order to provide the appropriate scaling of content displayed in a browser. This feature is still in development and therefore considered experimental.
```

fedejeanne reviewed (dismissed):
It works pretty well, I only have some minor comments about naming/typos and one important detail about preserving the selected browser when the new preference is set to `false`.

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/IWorkbenchPreferenceConstants.java:
```diff
@@ -705,4 +705,13 @@ public interface IWorkbenchPreferenceConstants {
+	/**
+	 * <p>
+	 * Whether the UI updates on runtime during dpi change
+	 * </p>
```
It's not only about updating the UI, it's actually about adapting to DPI changes
```suggestion
	 * <p>
	 * Whether the UI adapts to DPI changes at runtime
	 * </p>
```
Please also flag this preference as "experimental" and state that it only affects Windows.

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/IWorkbenchPreferenceConstants.java:
```diff
@@ -705,4 +705,13 @@ public interface IWorkbenchPreferenceConstants {
+	 *
+	 * @since 3.133
+	 */
+	String UPDATE_ON_RUNTIME_KEY = "UPDATE_ON_RUNTIME_KEY"; //$NON-NLS-1$
```
`UPDATE_ON_RUNTIME` is very generic and might conflict to other things. Could we use something more specific like at least `RESCALE_ON_RUNTIME` or `ADAPT_DPI_AT_RUNTIME` or the like?

fedejeanne replied:
I'd go with `RESCALING_AT_RUNTIME` since it matches the **public** (API) method `Display::setRescalingAtRuntime` and would adapt the names of other methods, parameters, constants, fields and information texts in this PR too in order to keep it consistent.

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/Workbench.java:
```diff
@@ -678,6 +678,15 @@ public void update() {
 	}
 
+	private static void setRuntimePropertyFromPreference(final Display display) {
+		boolean updateOnRunTime = PrefUtil.getAPIPreferenceStore()
```
Please use consistent way of writing "runtime" --> `updateOnRuntime`.
And please use more descriptive name (see previous comment)

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/Workbench.java:
```diff
@@ -678,6 +678,15 @@ public void update() {
 		return returnCode[0];
 	}
 
+	private static void setRuntimePropertyFromPreference(final Display display) {
```
This is a very generic method name for a very specific task. Shouldn't is be specific to processing the preference for rescaling?

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/Workbench.java:
```diff
@@ -678,6 +678,15 @@ public void update() {
+	private static void setRuntimePropertyFromPreference(final Display display) {
+		boolean updateOnRunTime = PrefUtil.getAPIPreferenceStore()
+				.getBoolean(IWorkbenchPreferenceConstants.UPDATE_ON_RUNTIME_KEY);
+		display.setRescalingAtRuntime(updateOnRunTime);
```
This changes existing behavior. Executing `display.setRescalingAtRuntime(false)` makes the application (more specifically the display thread) run in `System` DPI awareness mode while currently it just runs with whatever it was started with. So this must be put in the if statement as well.

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/WorkbenchMessages.java:
```diff
@@ -34,6 +34,14 @@ public class WorkbenchMessages extends NLS {
+
+	public static String RuntimeSettingChangeWarningTitle;
+
+	public static String RuntimeSettingChangeWarningDesc;
```
Please use comprehensible names (see previous comments). In particular "RuntimeSettingChangeWarningTitle" is rather unspecific.

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -192,6 +200,31 @@ protected Control createContents(Composite parent) {
+		Label infoLabel = new Label(group, SWT.WRAP);
+		infoLabel.setText(WorkbenchMessages.UpdateUIZoomDesc);
+		GridData labelGridData = new GridData(SWT.FILL, SWT.CENTER, true, false);
+		labelGridData.widthHint = 400; // Adjust this value to control initial wrapping width
```
Hardcoding this width value look inappropriate. Itsn't there an existing constant for this? Or any other way of making this horizontally fill the dialog, whatever its size may be?

Note that there is also `GridDataFactory` with some default to simplify the process of layout data generation.

fedejeanne replied:
That would be a 350

```java
infoLabel.setLayoutData(GridDataFactory.defaultsFor(infoLabel).create());
```

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -212,6 +245,9 @@ protected void createHideIconsForViewTabs(Composite composite) {
 				actualValue);
+		GridData gridData = new GridData();
+		gridData.horizontalIndent = 20;
+		hideIconsForViewTabs.setLayoutData(gridData);
```
Why it is necessary to now apply this grid data to `hideIconsForViewTabs` instead of `dependent` in the `createDependency` method like before?

ShahzaibIbrahim replied:
It's a leftover. I was trying something and later forgot to move it back.

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/messages.properties:
```diff
@@ -497,6 +497,10 @@ PreferencePageParameterValues_pageLabelSeparator = \ >\
 ThemeChangeWarningTitle = Theme Changed
+RuntimeSettingChangeWarningTitle = DPI Setting Changed
+RuntimeSettingChangeWarningDesc = Restart for the effect to take place
+UpdateUIZoomTitle = Update UI scaling on runtime
```
May it be more self-explaining when putting the `monitor-specific` property in the label? E.g.
```suggestion
UpdateUIZoomTitle = Monitor-specific UI scaling
```

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -341,6 +373,11 @@ public boolean performOk() {
+		boolean isUpdateOnRuntimeChanged = (updateOnRuntime != null
+				&& initialStateUpdateOnRuntime != updateOnRuntime.getSelection());
+		apiStore.setValue(IWorkbenchPreferenceConstants.UPDATE_ON_RUNTIME_KEY,
+				updateOnRuntime == null ? false : updateOnRuntime.getSelection());
```
Shouldn't this whole block be put into a `updateOnRuntime != null` condition? Why should the value be stored in the preference store if the button is not present (i.e., if we are not on Windows, if I understand correctly) at all?

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -373,13 +410,20 @@ public boolean performOk() {
+	}
+
+	private void showRestartDialog(String title, String warningText) {
+		if (new MessageDialog(null, title, null, warningText, MessageDialog.NONE, 2,
```
Please properly refactor this to either always have the consumer decise which title and text to put in the dialog or to have one parameterless method per use case. Now we have one method for custom title/text and one with some unspecified default values (which by accident default to themeing). I would propose to just have the method with two parameters and make the method consumers pass the required information.

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -114,6 +116,9 @@ public class ViewsPreferencePage extends PreferencePage implements IWorkbenchPre
 	private Button themingEnabled;
+	private Button updateOnRuntime;
+	private boolean initialStateUpdateOnRuntime = PrefUtil.getAPIPreferenceStore()
+			.getBoolean(IWorkbenchPreferenceConstants.UPDATE_ON_RUNTIME_KEY);
```
Why is this solved different then the themeing property does it? That one is simply retrieved from the preference store when identifying whether the value has changed instead of storing it in a field. In my opinion, that solution is better as this field just serves as an unnecessary and potentially error-prone cache.

HeikoKlare reviewed (changes requested):
Great to finally have a proper preference for enabling the enhanced HiDPI support on Windows, thanks!

Still, in the current state, this PR introduces a regression and needs some revision regarding design and naming. See my detailed comments.

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -127,6 +132,8 @@ protected Control createContents(Composite parent) {
 
 		themingEnabled = createCheckButton(comp, WorkbenchMessages.ThemingEnabled, engine != null);
+		initialStateUpdateOnRuntime = PrefUtil.getAPIPreferenceStore()
+				.getBoolean(IWorkbenchPreferenceConstants.RESCALING_AT_RUNTIME);
```
As mentioned in https://github.com/eclipse-platform/eclipse.platform.ui/pull/2461#discussion_r1824038278, I do not yet understand why we need to store this information in a field instead of retrieving it when required, like done for the themeing.

ShahzaibIbrahim replied:
So that it could be used across the class. I needed this field info while creating the HiDpi setting group and also when Ok is performed. For theming they checked the initial state by simply checking if theme != null. I hope it make sense

ShahzaibIbrahim replied:
Now, I am retrieving the value only in the methods where its needed, separately.

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -114,6 +117,8 @@ public class ViewsPreferencePage extends PreferencePage implements IWorkbenchPre
 	private boolean highContrastMode;
 
 	private Button themingEnabled;
+	private Button updateOnRuntime;
```
Can we call this `rescaleAtRuntime` to be consistent with the general naming? Same applies to other variables in this class.

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/messages.properties:
```diff
@@ -497,6 +497,11 @@ PreferencePageParameterValues_pageLabelSeparator = \ >\
+RescaleAtRuntimeSettingChangeWarningDesc = Restart for the effect to take place
+UpdateUIZoomGroupTitle = HiDPI settings
+UpdateUIZoomTitle = Monitor-specific UI scaling
+UpdateUIZoomDesc = EXPERIMENTAL! Activating this option will dynamically scale all windows according to the monitor they are currently in. It will also set the default browser to Edge in order to provide the appropriate scaling of content displayed in a browser. This feature is still in development and therefore considered experimental.
```
Can we please improve the property names here?
E.g.:
- `UpdateUIZoomGroupTitle` -> `HiDpiSettingsGroupTitle`
- `UpdateUIZoomTitle` -> `RescaleAtRuntimeEnabled`
- `UpdateUIZoomTitle` -> `RescaleAtRuntimeDisclaimer`

And maybe add a mnemonic for the new option? If I see correctly "s" is not used yet, so you might use `Monitor-specific UI &scaling`

HeikoKlare commented on bundles/org.eclipse.ui.workbench/Eclipse UI/org/eclipse/ui/internal/dialogs/ViewsPreferencePage.java:
```diff
@@ -192,6 +202,28 @@ protected Control createContents(Composite parent) {
 	}
 
+	private void createHiDPISettingsGroup(Composite parent) {
+		if (!OS.isWindows() || true) {
```
The options do not appear anymore because this always evaluates to `true`, such that the method is never executed.

ShahzaibIbrahim replied:
Sorry, leftover...

HeikoKlare reviewed (commented):
Regarding functionality, this looks good now (except for the case that the options do not show up anymore because of a faulty conditions).

Some minor code style comments are remaining.

ShahzaibIbrahim committed:
Adding experimental preference to enable rescaling at runtime (Windows)

In the UI Preferences -> General -> Appearance, there is a new HiDPI
setting that could be checked to enable the monitor-specific scaling of
the UI. This check will also enable Edge browser by default. This is an
experimental feature to test the multi-monitor HiDPI support before it
is actually released and set by default. For now, this preference is
disabled by default.

HeikoKlare reviewed (approved):
From my side, this also looks good now. Thank you!

HeikoKlare commented:
Failing test is unrelated/already documented: #370
